controllers/topology/internal/mergepatch/managed_paths.go

```diff
@@ -0,0 +1,138 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mergepatch
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	managedPathSeparator  = "."
+	managedFieldSeparator = ","
+)
+
+// DeepCopyWithManagedFieldAnnotation returns a copy of the object with an annotation
+// Keeping track of the fields the object is setting.
+func DeepCopyWithManagedFieldAnnotation(obj client.Object) (client.Object, error) {
+	// Store the list of paths managed by the topology controller in the current patch operation;
+	// this information will be used by the next patch operation.
+	objWithManagedFieldAnnotation := obj.DeepCopyObject().(client.Object)
+	if err := storeManagedPaths(objWithManagedFieldAnnotation); err != nil {
+		return nil, err
+	}
+	return objWithManagedFieldAnnotation, nil
+}
+
+// getManagedPaths infers the list of paths managed by the topology controller in the previous patch operation
+// by parsing the value of the managed field annotation.
+// NOTE: if for any reason the annotation is missing, the patch helper will fall back on standard
+// two-way merge behavior.
+func getManagedPaths(obj client.Object) []contract.Path {
+	// Gets the managed field annotation from the object.
+	managedFieldAnnotation := obj.GetAnnotations()[clusterv1.ClusterTopologyManagedFieldsAnnotation]
+
+	// Parses the managed field annotation value into a list of paths.
+	// NOTE: we are prepending "spec" to the paths to restore the correct absolute path inside the object.
+	managedFields := strings.Split(managedFieldAnnotation, managedFieldSeparator)
+	paths := make([]contract.Path, 0, len(managedFields))
+	for _, managedField := range managedFields {
+		if strings.TrimSpace(managedField) == "" {
+			continue
+		}
+
+		path := contract.Path{"spec"}
+		for _, f := range strings.Split(managedField, managedPathSeparator) {
+			path = append(path, strings.TrimSpace(f))
+		}
+		paths = append(paths, path)
+	}
+	return paths
+}
+
+// Stores the list of paths managed by the topology controller into the managed field annotation.
+// NOTE: The topology controller is only concerned about managed paths in the spec; given that
+// we are dropping spec. from the result to reduce verbosity of the generated annotation.
+// NOTE: Managed paths are relevant only for unstructured objects where it is not possible
+// to easily discover which fields have been set by templates + patches/variables at a given reconcile;
+// instead, it is not necessary to store managed paths for typed objets (e.g. Cluster, MachineDeployments)
+// given that the topology controller explicitly sets a well-known, immutable list of fields at every reconcile.
+func storeManagedPaths(obj client.Object) error {
```

Comment:
Currently we ignore `ignoredPaths` when storing managedPaths. This means the following can happen:
DockerClusterTemplate
```yaml
  spec:
    template:
      metadata: {}
      spec:
        controlPlaneEndpoint:
          host: ""
          port: 0
        loadBalancer: {}
```

DockerCluster
```yaml
  kind: DockerCluster
  metadata:
    annotations:
      cluster.x-k8s.io/cloned-from-groupkind: DockerClusterTemplate.infrastructure.cluster.x-k8s.io
      cluster.x-k8s.io/cloned-from-name: quick-start-my-cluster
      topology.cluster.x-k8s.io/managed-fields: controlPlaneEndpoint.host, controlPlaneEndpoint.port,
        loadBalancer.imageRepository, loadBalancer.imageTag
```

I think we should make sure that fields which are part of ignoredPaths should not end up in the managed fields annotation. I think the same applies for fields which are **not** part of allowedPaths (essentially the managed paths should reflect the filtering we're doing when patching with allowedPaths/ignorePaths)

Reply:
Now we are respecting IngorePaths. We are already implicitly respecting ignore paths (they are hard coded, and we are considering only one of them - spec)